### ISSUE TITLE
Confusing error when populating a struct with a field called 'start' without an explicit value and a variable called start is not defined

### ISSUE THREAD

**Issue by jjl** [rust-lang/rust#86290]
Given the following code:

```rust
struct Wat {
    start: (),
}
impl Wat {
    fn new() -> Wat {
      Wat { start }
    }
}
```

The current output is:

```
expected value, found built-in attribute `start`
a field by this name exists in `Self` [E0423]
```

I frankly have no idea what it's on about and I was surprised that it doesn't just tell me it can't find the variable of that name.

**Comment by mbartlett21** [comment]
This is because there is a builtin attribute `#[start]` that can be applied to a function to make it the entry point, which is what it resolves `start` to, since you have no variable shadowing it.

**Comment by jjl** [comment]
Ahh, gotcha. I don't think it's helpful to note that it found an attribute of that name, or at least it confused me.

**Comment by kpreid** [comment]
If the field name does not happen to be an attribute, the note is less confusing but still unhelpful:

```rust
struct Foo {
    bar: i32,
}

impl Foo {
    fn new() -> Self {
        Self {
            bar,
        }
    }
}
```

```
error[E0425]: cannot find value `bar` in this scope
 --> src/lib.rs:8:13
  |
8 |             bar,
  |             ^^^ a field by this name exists in `Self`

For more information about this error, try `rustc --explain E0425`.
error: could not compile `playground` due to previous error
```

It seems to me that there is little point in saying “a field by this name exists”, since we're in a struct initializer — if there _wasn't_ such a field, we'd be getting a different error. Or perhaps it would be helpful to say “a field by this name exists in `Self`, but there is no variable by this name to initialize it with; try writing `bar: something_else_here`”.

**Comment by AnthonyMikh** [comment]
Hit it today with a code that can be minified to this:

```rust
fn () {
    path
}
```

Giving

```
error[[E0423]](https://doc.rust-lang.org/stable/error-index.html#E0423): expected value, found built-in attribute `path`
 --> src/lib.rs:2:5
  |
2 |     path
  |     ^^^^ not a value
```

In my case it was even more confusing because I was rearranging my code and I actually had a variable named `path`, just in a nested scope.

Interestingly enough, Rust prior to version 1.38.0 gives more sensible error. For example, [Rust 1.37.0 gives the following](https://godbolt.org/z/1q6hhY1Wa):

```
error[E0425]: cannot find value `path` in this scope
 --> <source>:2:5
  |
2 |     path
  |     ^^^^ not found in this scope
```

**Comment by clarfonthey** [comment]
Want to call out that this is still happening in 2026: https://github.com/rust-lang/rust/pull/158168#issuecomment-4756832870

**Comment by petrochenkov** [comment]
The logic of the diagnostic is that if we cannot find a name in its expected namespace, then we try to find it in the 2 remaining namespaces.

We could also get something like "expected value, found enum `E`" or "expected value, found macro `format`" here.
Why are built-in attributes specifically problematic in this case? How are they different from the cases above?
I don't think we should change any logic here to special-case built-in attributes.
Maybe only tweak the wording to make it clear that we are searching for the same name in different namespaces.

**Comment by TorinAsakura** [comment]
@petrochenkov and I concluded that this is worth discussing a bit more broadly:

A possible wording for cross-namespace cases:

- expected value, but a built-in attribute with the same name exists in another namespace
- expected value, but an enum with the same name exists in another namespace
- expected type, but a constant with the same name exists in another namespace

Same-namespace diagnostics, error codes, labels, suggestions, and resolution order would stay unchanged. The diagnostic branch would be selected when res.ns() differs from source.namespace().

Does this wording look reasonable?

**Comment by clarfonthey** [comment]
I guess that part of the pause with respect to attributes is that it feels completely reasonable to accidentally use a type name, but since the attribute syntax requires so much extra wrapping, the attribute namespace feels totally unrelated. I would expect it to still say something along the lines of "name not defined (as a variable)" even if it mentions it found the attribute.

**Comment by TorinAsakura** [comment]
@clarfonthey Yeah, that distinction makes sense. The missing variable should stay front and center, and the attribute should only show up as extra context. Something like:

- no value named path is defined in this scope
- note: a built-in attribute named path exists in another namespace

Then we can use the same shape for the other cross-namespace cases. Is that closer to what you had in mind?

**Comment by clarfonthey** [comment]
Yes, except there should also be the usual "did you forget to define it?" caveat and the presence in another namespace shouldn't remove it.

**Comment by TorinAsakura** [comment]
@clarfonthey Got it. Then the cross-namespace lookup should be purely additive: keep the ordinary missing-name diagnostic and all of its suggestions, and only add a note about the name found in another namespace.
